internal/runtime/storeforward.go

```diff
@@ -0,0 +1,188 @@
+//
+// Copyright (c) 2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package runtime
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"runtime"
+	"time"
+
+	"github.com/edgexfoundry/app-functions-sdk-go/appcontext"
+	"github.com/edgexfoundry/app-functions-sdk-go/internal/common"
+	"github.com/edgexfoundry/app-functions-sdk-go/internal/store/contracts"
+	"github.com/edgexfoundry/app-functions-sdk-go/internal/store/db/interfaces"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients"
+)
+
+const (
+	DefaultMinRetryInterval = 5000 // 5 secs
+)
+
+type storeForwardInfo struct {
+	runtime      *GolangRuntime
+	storeClient  interfaces.StoreClient
+	pipelineHash string
+}
+
+func (sf *storeForwardInfo) startStoreAndForwardRetryLoop(appCtx context.Context, enabledCtx context.Context, serviceKey string, config *common.ConfigurationStruct, edgeXClients common.EdgeXClients) {
+
+	go func() {
```

Comment:
While you're using cancelable contexts, I don't see anything in your implementation that waits for the cancelled threads to actually stop executing.

Reply:
That will be in the next PR where the context is created and passed in.

Reply:
Actually, I don't have the required wait group... Will add.

Reply:
Added the needed wait groups...